### ISSUE TITLE
[minillm]cannot reproduce the results gpt2 1.5B -> 120M

### ISSUE THREAD

**Issue by 01yingxuan** [microsoft/LMOps#314]
This is my results(8*L20, 40G):

Dollyeval | SelfInst | VicunaEval | S-NI | UnNI
-- | -- | -- | -- | --
24.6 | 13.2 | 16.9 | 25.3 | 26.6
24.1489 | 9.7044 | 16.4412 | 16.1367 | 18.0202

Can you help figure out what's wrong with my training scripts?
_```
#! /bin/bash
export CUDA_VISIBLE_DEVICESi=0,1,2,3,4,5,6,7
MASTER_ADDR=localhost
MASTER_PORT=${2-2012}
NNODES=1
NODE_RANK=0
GPUS_PER_NODE=${3-8}

DISTRIBUTED_ARGS="--nproc_per_node $GPUS_PER_NODE \
                  --nnodes $NNODES \
                  --node_rank $NODE_RANK \
                  --master_addr $MASTER_ADDR \
                  --master_port $MASTER_PORT"

# model
BASE_PATH=${1-"/workspace/micro/minillm"}
CKPT_NAME="sft/e20-bs2-lr0.0005-G1-N8-NN1/13566"
CKPT="${BASE_PATH}/results/gpt2/train/${CKPT_NAME}/"
TEACHER_CKPT_NAME="xlarge-sft"
TEACHER_CKPT="${BASE_PATH}/results/gpt2/train/sft/e10-bs2-lr5e-05-G1-N8-NN1/6426/"

# data
PROMPT_DATA_DIR="${BASE_PATH}/processed_data/dolly/prompt/gpt2/"
LM_DATA_DIR="${BASE_PATH}/processed_data/openwebtext/gpt2/512/10M/"
# runtime
SAVE_PATH="${BASE_PATH}/results/gpt2/train/minillm/0508"
# hp
GRAD_ACC=1
BATCH_SIZE=16 # 16
CHUNK_SIZE=16
OPTS=""
# model
OPTS+=" --base-path ${BASE_PATH}"
OPTS+=" --model-path ${CKPT}"
OPTS+=" --teacher-model-path ${TEACHER_CKPT}"
OPTS+=" --ckpt-name ${CKPT_NAME}"
OPTS+=" --teacher-ckpt-name ${TEACHER_CKPT_NAME}"
OPTS+=" --n-gpu ${GPUS_PER_NODE}"
OPTS+=" --n-nodes ${NNODES}"
OPTS+=" --teacher-model-fp16" # fp16
# OPTS+=" --gradient-checkpointing"
# data
OPTS+=" --prompt-data-dir ${PROMPT_DATA_DIR}"
OPTS+=" --lm-data-dir ${LM_DATA_DIR}"
OPTS+=" --dev-num 1000"
OPTS+=" --num-workers 0"
# hp
OPTS+=" --epochs 40" # 10
OPTS+=" --total-iters 10000" # 5000
OPTS+=" --kd-ratio 0.5"
OPTS+=" --batch-size ${BATCH_SIZE}"
OPTS+=" --lr 2.5e-6" #5e-6
OPTS+=" --lr-min 2.5e-6" #5e-6
OPTS+=" --gradient-accumulation-steps ${GRAD_ACC}"
OPTS+=" --max-length 512"
OPTS+=" --max-prompt-length 256"
OPTS+=" --warmup-iters 100"
# runtime
OPTS+=" --save ${SAVE_PATH}"
OPTS+=" --seed 10"
OPTS+=" --seed-ppo 42"
OPTS+=" --seed-lm 7"
OPTS+=" --save-interval 500"
OPTS+=" --eval-interval 100"
OPTS+=" --log-interval 16"
OPTS+=" --mid-log-num 1"
# ppo
OPTS+=" --type minillm"
OPTS+=" --ppo-epochs 4"
OPTS+=" --num-rollouts 256"
OPTS+=" --chunk-size ${CHUNK_SIZE}"
# minillm
OPTS+=" --length-norm"
OPTS+=" --single-step-reg"
OPTS+=" --teacher-mixed-alpha 0.2"
# reward
OPTS+=" --reward-scaling 0.5"
OPTS+=" --cliprange-reward 100"
# gen
OPTS+=" --do-sample"
OPTS+=" --top-k 0"
OPTS+=" --top-p 1.0"
OPTS+=" --temperature 1.0"
# deepspeed
OPTS+=" --deepspeed"
OPTS+=" --deepspeed_config ${BASE_PATH}/configs/deepspeed/ds_config_zero1_fp16.json"

export NCCL_DEBUG=""
export WANDB_DISABLED=True
export TF_CPP_MIN_LOG_LEVEL=3
export PYTHONPATH=${BASE_PATH}
CMD="torchrun ${DISTRIBUTED_ARGS} ${BASE_PATH}/train_minillm.py ${OPTS} $@"

echo ${CMD}
echo "PYTHONPATH=${PYTHONPATH}"
mkdir -p ${SAVE_PATH}
${CMD}
```_


p.s. my SFT results seems right:

Method | Dollyeval | SelfInst | VicunaEval | S-NI | UnNI
-- | -- | -- | -- | -- | --
SFT w/o KD | 23.3 | 10 | 14.7 | 16.3 | 18.5
SFT w/o KD | 22.6223 | 9.2331 | 14.7403 | 15.3088 | 17.3898

**Comment by 01yingxuan** [comment]
![Image](https://github.com/user-attachments/assets/274cd916-4e37-4fbd-9547-74fe4461467f)here is my loss curve

**Comment by t1101675** [comment]
It seems that the initial checkpoint `CKPT_NAME="sft/e20-bs2-lr0.0005-G1-N8-NN1/13566"` is not selected based on the lowest validation loss. The model reaches the lowest validation loss at around 1K steps. You can try using the checkpoint with the lowest validation loss (as in Sec. B.1 in our paper), or directly use [MiniLLM/init-gpt2-120M](https://huggingface.co/MiniLLM/init-gpt2-120M) as the initialization.

Besides, we use a total batch size of 256 (GPUS_PER_NODE * BATCH_SIZE). Try increasing this.

**Comment by 01yingxuan** [comment]
> It seems that the initial checkpoint `CKPT_NAME="sft/e20-bs2-lr0.0005-G1-N8-NN1/13566"` is not selected based on the lowest validation loss. The model reaches the lowest validation loss at around 1K steps. You can try using the checkpoint with the lowest validation loss (as in Sec. B.1 in our paper), or directly use [MiniLLM/init-gpt2-120M](https://huggingface.co/MiniLLM/init-gpt2-120M) as the initialization.初始检查点 `CKPT_NAME=“sft/e20-bs2-lr0.0005-G1-N8-NN1/13566”` 似乎不是根据最低的验证损失来选择的。该模型在大约 1K 步长处达到最低的验证损失。您可以尝试使用验证损失最低的检查点（如我们论文中的第 B.1 节），或者直接使用 [MiniLLM/init-gpt2-120M](https://huggingface.co/MiniLLM/init-gpt2-120M) 作为初始化。
> 
> Besides, we use a total batch size of 256 (GPUS_PER_NODE * BATCH_SIZE). Try increasing this.此外，我们使用的总批量大小为 256 （GPUS_PER_NODE * BATCH_SIZE）。尝试增加此值。

Thank you for replying so quickly ! I will give the method you mentioned a try.
